frontend/catalyst/pennylane_extensions.py

```diff
@@ -395,15 +400,27 @@ def __call__(self, *args, **kwargs):
 
         if EvaluationContext.is_tracing():
             fn = _ensure_differentiable(self.fn)
+            argnum_is_int = (
+                isinstance(self.grad_params.argnum, int) or self.grad_params.argnum is None
+            )
             grad_params = _check_grad_params(*self.grad_params)
             jaxpr, out_tree = _make_jaxpr_check_differentiable(fn, grad_params, *args)
 
-            args_data, _ = tree_flatten(args)
+            args_data, in_tree = tree_flatten(args)
+
+            total_argnums = list(range(0, len(args_data)))
+            argnums_unflatten = tree_unflatten(in_tree, total_argnums)
+            argnums_selected = [argnums_unflatten[i] for i in grad_params.argnum]
+            argnums_correct, _ = tree_flatten(argnums_selected)
+            args_argnum = tuple(args[i] for i in grad_params.argnum)
+            _, in_tree = tree_flatten(args_argnum)
```

Comment:
Regarding your question, this seems like a decent solution to me. One alternative I can think of would be:
```py
flat_argnums = []
current_idx = 0
for i, arg in enumerate(args):
    num_leaves = len(jax.tree_util.leaves(arg))
    if i in grad_params.argnum:
        flat_argnums.extend(range(current_idx, current_idx + num_leaves))
    current_idx += num_leaves
```
I'll leave it to you if like this more or the flatten/unflatten method.

Reply:
Thanks, I think I will just clean my implementation